src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -318,20 +351,19 @@ TransactionManager::do_submit_transaction(
   LOG_PREFIX(TransactionManager::do_submit_transaction);
   SUBTRACET(seastore_t, "start", tref);
   return trans_intr::make_interruptible(
-    tref.get_handle().enter(write_pipeline.ool_writes)
-  ).then_interruptible([this, FNAME, &tref,
+    tref.get_handle().enter(write_pipeline.ool_writes_and_lba_updates)
+  ).then_interruptible([this, &tref,
 			dispatch_result = std::move(dispatch_result)] {
     return seastar::do_with(std::move(dispatch_result),
-			    [this, FNAME, &tref](auto &dispatch_result) {
+			    [this, &tref](auto &dispatch_result) {
       return epm->write_delayed_ool_extents(tref, dispatch_result.alloc_map
-      ).si_then([this, FNAME, &tref, &dispatch_result] {
-        SUBTRACET(seastore_t, "update delayed extent mappings", tref);
-        return lba_manager->update_mappings(tref, dispatch_result.delayed_extents);
-      }).handle_error_interruptible(
+      ).handle_error_interruptible(
         crimson::ct_error::input_output_error::pass_further(),
         crimson::ct_error::assert_all("invalid error")
       );
     });
+  }).si_then([&tref, this] {
+    return update_lba_mappings(tref);
```

Comment:
Can `update_lba_mappings()` be moved to line 376 below?

I suppose the special version of `Trans::for_each_fresh_block_pre_submit()` will be no longer necessary, which looks confusing vs `Trans::for_each_fresh_block()`.

Reply:
> Can `update_lba_mappings()` be moved to line 376 below?
> 

Um, it doesn't seem feasible to me, because `update_lba_mappings()` are actually updating on-disk checksums which would be written to disk by `epm->write_preallocated_ool_extents` at line 371, so it has to be done before line 371.

> I suppose the special version of `Trans::for_each_fresh_block_pre_submit()` will be no longer necessary, which looks confusing vs `Trans::for_each_fresh_block()`.

Reply:
Ah, I see.

Is this way more explict?:
1. Rename `Trans::for_each_fresh_block()` to a more explicit name `Trans::for_each_finalized_fresh_block()`.
2. Firstly update mappings to the already "finalized" fresh extents, with a comment that it is for crc.
3. Update mappings after writing the delayed ool extents, with a comment that it is for both crc and paddr.
4. Update mappings before writing the preallocated ool extents, with a comment that it is for crc.

And I suppose we can parallelize writing delayed and preallocated ool extents to disk, and this will also be helpful.

Does it make sense?

Reply:
> Ah, I see.
> 
> Is this way more explict?:
> 
> 1. Rename `Trans::for_each_fresh_block()` to a more explicit name `Trans::for_each_finalized_fresh_block()`.
> 2. Firstly update mappings to the already "finalized" fresh extents, with a comment that it is for crc.

I'm a little confused, there's no finalized fresh extents before writing delayed or prealloc'd ool extents, am I missing right? Or am I missing something?

> 3. Update mappings after writing the delayed ool extents, with a comment that it is for both crc and paddr.
> 4. Update mappings before writing the preallocated ool extents, with a comment that it is for crc.
> 
> And I suppose we can parallelize writing delayed and preallocated ool extents to disk, and this will also be helpful.

Do you mean that we parallelize them in this way:

Fiber 1:

1. write delayed ool extents
2. update lba mappings for the delayed ool extents

Fiber 2:

1. update lba mappings for prealloc'd ool extents
2. write prealloc'd ool extents


> 
> Does it make sense?

Reply:
> I'm a little confused, there's no finalized fresh extents before writing delayed or prealloc'd ool extents, am I right? Or am I missing something?

Sorry I might not have chosen the most appropriate word.

Upon entering the phase `ool_writes_and_lba_updates`, some of the fresh extents are already added in `inline_block_list`, the other fresh extents can only fall into 2 categories:
* case 2: in `delayed_alloc_list`
* case 3: in `pre_alloc_list`

Case 4 (inplace-rewrite extents) is special because I don't think they should belong to the "fresh" extent category.

Their constraints and requirements to update the lba mappings are different.

> Do you mean that we parallelize them in this way.

Parallelization is unrelated to this PR. But we should still make sure the logic here is appropriately reflecting the inherent constraints and intentions.

Reusing the existing `Trans::for_each_fresh_block()` upon entering `ool_writes_and_lba_updates` might be appropriate for the case 1, but I'm not able to find a better word than "finalized".

Reply:
> Parallelization is unrelated to this PR. But we should still make sure the logic here is appropriately reflecting the inherent constraints and intentions.

For example, rearrage the logic to update the lba mappings for the below scenarios independently (quote from https://github.com/ceph/ceph/pull/55449#discussion_r1508693739):
1. Update mappings to the already "finalized" fresh extents, with a comment that it is for crc.
2. Update mappings after writing the delayed ool extents, with a comment that it is for both crc and paddr.
3. Update mappings before writing the preallocated (and inplace-rewrite) ool extents, with a comment that it is for crc.

Reply:
> Reusing the existing `Trans::for_each_fresh_block()` upon entering `ool_writes_and_lba_updates` might be appropriate for the case 1, but I'm not able to find a better word than "finalized".

I think the fresh extents' (lba/backref in most cases) contents might be changed during ool extents write, so we have calculate their CRCes after the ool write phase, am I right?

Reply:
We need to identify the dependencies about the 3 cases (https://github.com/ceph/ceph/pull/55449#discussion_r1510743480) with lba-mapping and crc updates.

> I think the fresh extents' (lba/backref in most cases) contents might be changed during ool extents write, so we have calculate their CRCes after the ool write phase, am I right?

The related decisions should have been encapsulated within `EPM::alloc_new_(non_)data_extent()`. If you are referring to case 2 (delayed-ool fresh extents), IIUC it should be impossible for lba/backref to be in the case 2.

Also, we cannot generally adjust lba/backref's crc after they are written, correct?

Reply:
> We need to identify the dependencies about the 3 cases ([#55449 (comment)](https://github.com/ceph/ceph/pull/55449#discussion_r1510743480)) with lba-mapping and crc updates.
> 
> > I think the fresh extents' (lba/backref in most cases) contents might be changed during ool extents write, so we have calculate their CRCes after the ool write phase, am I right?
> 
> The related decisions should have been encapsulated within `EPM::alloc_new_(non_)data_extent()`. ~If you are referring to case 2 (delayed-ool fresh extents), IIUC it should be impossible for lba/backref to be in the case 2.~ Seems you are referring to the case 1.
> 
> ~But we should not generally adjust lba/backref's crc after they are written, especially with the sequential/segmented backend, correct?~

Yeah, I meant case 1, the physical addresses of logical extents are determined after the ool write phase, which means the lba leaf nodes' entries may be changed after the ool write phase, so I think we may have to delay the crc calculation of lba/bacref extents after the ool write phase and before the pre-alloc write phase.

Reply:
Got it!

I think the name `for_each_pre_submit_fresh_block` is misleading: inplace-rewrite extents are not fresh, and pre_submit is still vague.

However, seems at the time after delayed-ool-write and before preallocate-ool-write, `for_each_pre_submit_fresh_block` is identical to `for_each_finalized_fresh_block` + `get_valid_pre_alloc_list()`. Do you think it's reasonable to adjust the logic to reuse them properly?

Meanwhile, the following comments might be important to add and remember, feel free to adjust them to be more precise:
```c++
// Update both crc and paddr in the mappings for the written delayed-ool fresh extents, which must be logical.
// Update crc in the mappings for the rest of fresh or inplace-rewrite extents, before going to write any lba/backref extents.
```